File: Python/047 Permutations II.py
# Given a collection of numbers that might contain duplicates, return all possible unique permutations.

# Example:

# Input: [1,1,2]
# Output:
# [
#   [1,1,2],
#   [1,2,1],
#   [2,1,1]
# ]


class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        length = len(nums)
        result = []

        def callback(i, group):
            if len(group) == length:
                arr = [nums[i] for i in group]
                result.append(arr)
                return
            for j in range(0, length):

                if j in group:
                    continue
                if j > 0 and nums[j] == nums[j - 1] and j - 1 not in group:
                    continue
                group.append(j)
                callback(j, group)
                group.pop()

        for i in range(0, length):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            callback(i, [i])
        return result

# test
s = Solution()
print(s.permuteUnique([1, 1, 2]))
print(s.permuteUnique([3, 3, 0, 3]))

File: JavaScript/047 Permutations II.js
// Given a collection of numbers that might contain duplicates, return all possible unique permutations.

// Example:

// Input: [1,1,2]
// Output:
// [
//   [1,1,2],
//   [1,2,1],
//   [2,1,1]
// ]


/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function (nums) {
    let result = [];
    let len = nums.length;
    nums = nums.sort(function (i, j) {
        return i - j;
    })
    let callback = function (i, group) {
        if (group.length === len) {
            result.push(group.map(function (i) {
                return nums[i];
            }));
            return;
        }
        for (let j = 0; j < len; j++) {
            if (group.indexOf(j) !== -1) continue;
            if (j > 0 && nums[j] === nums[j - 1] && group.indexOf(j - 1) === -1) {
                continue;
            }
            group.push(j);
            callback(j, group);
            group.pop();
        }
    }
    for (let i = 0; i < len; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;
        }
        callback(i, [i])
    }
    return result;
};

// test
console.log(permuteUnique([1, 1, 2]));
console.log(permuteUnique([3,3,0,3]));